core/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -354,6 +354,12 @@ public void triggerVariableListenersInNotificationQueues() {
         }
         if (shadowVariableSession != null) {
             shadowVariableSession.updateVariables();
+            // Some internal variable listeners (such as those used
+            // to check for solution corruption) might have a declarative
+            // shadow variable as a source and need to be triggered here.
+            for (var notifiable : notifiableRegistry.getAll()) {
+                notifiable.triggerAllNotifications();
+            }
```

Comment:
This code is becoming very confusing - we trigger, then the cascade, then we update shadows again, and then we trigger again. 

We need to start simplifying it - cascade can easily be converted to DSV. And eventually all the other listeners too. Something to look forward to.

Reply:
This is strictly for our own variable listeners; user variable listeners cannot have declarative variable listeners as a source, but our own internal ones might. In particular, the variable listeners that track the before/after variable changed calls to check for solution corruption.